### ISSUE TITLE
Training the movie+mcan model using the default config cannot reproduce the reported test-dev score on VQA.

### ISSUE THREAD

**Issue by yangapku** [facebookresearch/mmf#626]
## ❓ Questions and Help
Hi, I have trained a movie+mcan model on my own server using the default config yaml (batchsize=32, update_frequency=2, step=236000). Now I have many checkpoints which are saved every 1k steps. May I ask which one among them should be picked for inference and achieve better performance compared with others according to your experience? Will the movie_mcan_final.pth just be okay? Thank you very much!

**Comment by vedanuj** [comment]
For the movie+mcan submission to VQA Challenge we picked the last checkpoint. You can also use the `movie_mcan_final.pth`.

**Comment by yangapku** [comment]
@vedanuj  Hi, I used the generated movie_mcan_final.pth checkpoint for inference. However, the test-dev score is much lower than expected (test-dev "yes/no": 86.96, "number": 56.23, "other": 62.52, "overall": 71.87). Can you help to figure out where the problem is? Thank you!

**Comment by yangapku** [comment]
The training config: [link](https://yangan.oss-cn-beijing.aliyuncs.com/mmf_checkpoint/rerun_bs32_uf2_fixbug_seed16304007-16310113/config.yaml).

**Comment by yangapku** [comment]
I also tried the last checkpoint (236000 steps), the performance is slightly better (test-dev yes/no: 87.16, number: 56.78, other: 63.01, overall: 72.25) but still not satisfying. Are there some implementation details I missed? Thank you very much!

**Comment by vedanuj** [comment]
You are using batch_size=32 and update_frequency=2 and step=236000. You are essentially doing half the number of updates to the model since the max_updates=236000 which is kept constant as used with batch_size=64 in our original runs. You will need to double the number of steps or double your batch_size.

**Comment by yangapku** [comment]
Thank you very much! Since the 16GB memory of my GPUs cannot hold batch_size=64, I think I should try to double the max_updates.

**Comment by yangapku** [comment]
@vedanuj Hi, I noticed that in the implementation codes, it seems that the `num_updates` increases 1 when the batch passed `update_frequency` times. So does it mean the model essentially performed `max_updates` times of updates with "real batch size" `batch_size * update_frequency` rather than half of that?

**Comment by vedanuj** [comment]
Sorry my bad! You are right. The number of epochs should be same irrespective of `update_frequency`. I will have to retrain and check why it is not giving the same result.

**Comment by yangapku** [comment]
Okay, thank you very much! Look forward to the solution.

**Comment by yangapku** [comment]
Hi, I notice that in the pull request #604, the `padded_feat` is actually created in the `read()` method of the reader but unused, where the unpadded `feat` is directly returned. Applying the fix during inference indeed helped us to reproduce the result of the provided checkpoint in the model zoo, but will it cause some problems if applied in training? I'm not very sure.

**Comment by yangapku** [comment]
@vedanuj @apsdehal  Hi, may I ask whether there is any progress on this issue? Thank you!

**Comment by vedanuj** [comment]
I have started training the model. Once the training completes then we will debug if there is any issue.

**Comment by yangapku** [comment]
@vedanuj @apsdehal  Hi, may I ask whether the training of the model has beed finished? Is the reported test-dev/test-std score reproduced successfully? Thank you very much!

**Comment by vedanuj** [comment]
@yangapku We finished training a model and we can get VQA test-dev score of 73.7. I am not yet sure if using the `update_frequency=2` is causing some type of regression in the final score. We are still investigating.

**Comment by yangapku** [comment]
@vedanuj Thank you for the reply! May I ask the commit and branch information of the repo you use?

**Comment by vedanuj** [comment]
I am using the master branch and on commit 97f3c7d375881d73a6465ec8cbb7e0ad2a7230a7

**Comment by yangapku** [comment]
@vedanuj Hi, may I ask whether there is any progress on checking if using `update_frequency=2`, `batch_size=32` and `max_updates=236000` will cause some type of regression in the final score? Thank you very much!

**Comment by yangapku** [comment]
@vedanuj @apsdehal Hi, sorry to bother again. Are there any following ups about the issue of `update_frequency `?

**Comment by vedanuj** [comment]
I am testing the model with update_frequency=2. Will respond here once the model finishes training.

**Comment by yangapku** [comment]
Okay! Thank you very much! 👍

**Comment by yangapku** [comment]
@vedanuj @apsdehal Hi, are there any evaluation results of training the model with `update_frequency=2`?

**Comment by vedanuj** [comment]
Hey @yangapku,  I saw similar drop in performance with `update_frequency=2`. I got VQA accuracy around 72.73. There is an issue with gradient accumulation. 

For now I recommend you use batch size 64 with same learning rate OR try lower learning rates with batch size 32. I will debug the issue with `update_frequency` but expect some delay in that.

**Comment by yangapku** [comment]
Okay, thank you for your following up and suggestion! I will try to use batch size 64. Hope the issue can be tackled soon.
